tools/build_node.js

```diff
@@ -18,30 +46,69 @@ async function buildNodeIndex(languages) {
     return `hljs.registerLanguage('${lang.name}', ${require});`;
   });
 
-  // legacy
-  await fs.writeFile(`${process.env.BUILD_DIR}/lib/highlight.js`,
-    "// This file has been deprecated in favor of core.js\n" +
-    "var hljs = require('./core');\n"
-  );
-
   const index = `${header}\n\n${registration.join("\n")}\n\n${footer}`;
-  await fs.writeFile(`${process.env.BUILD_DIR}/lib/index.js`, index);
+  await fs.writeFile(`${process.env.BUILD_DIR}/lib/${name}.js`, index);
 }
 
 async function buildNodeLanguage(language) {
   const input = { ...config.rollup.node.input, input: language.path };
   const output = { ...config.rollup.node.output, file: `${process.env.BUILD_DIR}/lib/languages/${language.name}.js` };
   await rollupWrite(input, output);
+  await rollupWrite(input, {...output,
+    format: "es",
+    file: output.file.replace("/lib/", "/es/")
+  });
 }
 
+const EXCLUDE = ["join"];
+
+async function buildESMUtils() {
+  const input = { ...config.rollup.node.input, input: `src/lib/regex.js` };
+  input.plugins = [...input.plugins, {
+    transform: (code) => {
+      EXCLUDE.forEach((fn) => {
+        code = code.replace(`export function ${fn}(`, `function ${fn}(`);
+      });
+      return code;
+    }
+  }];
+  const output = {
+    ...config.rollup.node.output,
+    format: "es",
+    file: `${process.env.BUILD_DIR}/es/utils/regex.js`
+  };
+  await rollupWrite(input, output);
+}
+
+
 async function buildNodeHighlightJS() {
   const input = { ...config.rollup.node.input, input: `src/highlight.js` };
   const output = { ...config.rollup.node.output, file: `${process.env.BUILD_DIR}/lib/core.js` };
   await rollupWrite(input, output);
+  await rollupWrite(input, { ...output,
+    format: "es",
+    file: `${process.env.BUILD_DIR}/es/core.js`
+  });
+}
+
+function dual(file) {
+  return {
+    require: file,
+    import: file.replace("/lib/", "/es/")
+  };
 }
 
 async function buildPackageJSON() {
   const packageJson = require("../package");
+
+  const exports = {
+    ".": dual("./lib/index.js"),
+    "./lib/common": dual("./lib/common.js"),
+    "./lib/core": dual("./lib/core.js"),
+    "./lib/languages/*": dual("./lib/languages/*.js"),
+  };
+  packageJson.exports = exports;
```

Comment:
From the Node.js documentation

> Warning: Introducing the "exports" field prevents consumers of a package from using any entry points that are not defined, including the package.json (e.g. require('your-package/package.json'). **This will likely be a breaking change.**

Should we add `package.json` to the export map? I'm not using it myself, but I wanted to make sure this was considered before landing in case it is useful for someone else.

Reply:
It would just need a single entry to itself, not dual, right?

Reply:
Yes, there is only one JSON format thankfully 😅